File: day-15/rambunctious_recitation.py
import collections

input_file = 'day-15/input.txt'

def van_eck(input, iterations):
  given = [int(x) for x in input.split(',')]
  last_seen = collections.defaultdict(int)
  for i in range(len(given)-1):
    last_seen[given[i]] = i+1
    
  previous = given[-1]
  for i in range (len(given),iterations):
    last_seen[previous], previous = i, 0 if previous not in last_seen else i - last_seen[previous]
  return previous

def part1(input):
  return van_eck(input, 2020)

def part2(input):
  return van_eck(input, 30000000)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-15/rambunctious_recitation.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

const input_file string = "./day-15/input.txt"

func ReadInput(input string) []int {
	numbers := []int{}
	for _, numString := range strings.Split(strings.TrimSpace(input), ",") {
		num, _ := strconv.Atoi(numString)
		numbers = append(numbers, num)
	}
	return numbers
}

func VanEck(input string, iterations int) int {
	last_seen := make(map[int]int)
	previous := 0
	for i, previous := range ReadInput(input) {
		last_seen[previous] = i + 1
	}

	for i := len(last_seen) + 1; i < iterations; i++ {
		next := 0
		if last, found := last_seen[previous]; found {
			next = i - last
		}
		last_seen[previous] = i
		previous = next
	}
	return previous
}

func Part1(input string) int {
	return VanEck(input, 2020)
}

func Part2(input string) int {
	return VanEck(input, 30000000)
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
